perses/dispersed/feptasks.py

```diff
@@ -84,30 +110,97 @@ def _get_integrator(self, thermodynamic_state):
         """
         return self._integrator
 
-    def reset(self):
-        """
-        Reset the work statistics on the associated ContextCache integrator.
+    def apply(self, thermodynamic_state, sampler_state):
+        """Propagate the state through the integrator.
+        This updates the SamplerState after the integration. It will apply the full NCMC protocol.
 
         Parameters
         ----------
         thermodynamic_state : openmmtools.states.ThermodynamicState
-            the thermodynamic state for which this integrator is cached.
+           The thermodynamic state to use to propagate dynamics.
+        sampler_state : openmmtools.states.SamplerState
+           The sampler state to apply the move to. This is modified.
         """
-        self._integrator.reset()
+        """Propagate the state through the integrator.
+        This updates the SamplerState after the integration. It also logs
+        benchmarking information through the utils.Timer class.
+        Parameters
+        ----------
+        thermodynamic_state : openmmtools.states.ThermodynamicState
+           The thermodynamic state to use to propagate dynamics.
+        sampler_state : openmmtools.states.SamplerState
+           The sampler state to apply the move to. This is modified.
+        See Also
+        --------
+        openmmtools.utils.Timer
+        """
+        # Check if we have to use the global cache.
+        if self.context_cache is None:
+            context_cache = cache.global_context_cache
+        else:
+            context_cache = self.context_cache
+
+        # Create integrator.
+        integrator = self._get_integrator(thermodynamic_state)
+
+        context, integrator = context_cache.get_context(thermodynamic_state, integrator)
+
+        # Subclasses may implement _before_integration().
+        self._before_integration(context, thermodynamic_state)
+        
+        self._cumulative_work[0] = integrator.get_protocol_work()
```

Comment:
Shouldn't this be zero? Maybe this is a good place to put an `assert` to ensure that it is.

Reply:
It should be zero. You asked me to have the first item in the array be zero.

Reply:
Yes! I'm just suggesting you add an `assert`.

Reply:
I'm raising a `RuntimeError` with an appropriate message if it's not zero now.